pkg/app/piped/planpreview/handler.go

```diff
@@ -0,0 +1,228 @@
+// Copyright 2021 The PipeCD Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package planpreview
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"time"
+
+	"go.uber.org/zap"
+
+	"github.com/pipe-cd/pipe/pkg/config"
+	"github.com/pipe-cd/pipe/pkg/git"
+	"github.com/pipe-cd/pipe/pkg/model"
+)
+
+type options struct {
+	workerNum              int
+	commandQueueBufferSize int
+	commandCheckInterval   time.Duration
+	commandHandleTimeout   time.Duration
+	logger                 *zap.Logger
+}
+
+type Option func(*options)
+
+func WithWorkerNum(n int) Option {
+	return func(opts *options) {
+		opts.workerNum = n
+	}
+}
+
+func WithCommandQueueBufferSize(s int) Option {
+	return func(opts *options) {
+		opts.commandQueueBufferSize = s
+	}
+}
+
+func WithCommandCheckInterval(i time.Duration) Option {
+	return func(opts *options) {
+		opts.commandCheckInterval = i
+	}
+}
+
+func WithCommandHandleTimeout(t time.Duration) Option {
+	return func(opts *options) {
+		opts.commandHandleTimeout = t
+	}
+}
+
+func WithLogger(l *zap.Logger) Option {
+	return func(opts *options) {
+		opts.logger = l
+	}
+}
+
+type gitClient interface {
+	Clone(ctx context.Context, repoID, remote, branch, destination string) (git.Repo, error)
+}
+
+type applicationLister interface {
+	List() []*model.Application
+}
+
+type commandLister interface {
+	ListBuildPlanPreviewCommands() []model.ReportableCommand
+}
+
+type Handler struct {
+	gitClient     gitClient
+	commandLister commandLister
+
+	commandCh    chan model.ReportableCommand
+	prevCommands map[string]struct{}
+
+	options        *options
+	builderFactory func() Builder
+	logger         *zap.Logger
+}
+
+func NewHandler(gc gitClient, cl commandLister, al applicationLister, cfg *config.PipedSpec, opts ...Option) *Handler {
+	opt := &options{
+		workerNum:              3,
+		commandQueueBufferSize: 10,
+		commandCheckInterval:   5 * time.Second,
+		commandHandleTimeout:   5 * time.Minute,
```

Comment:
nit: they should be defined as const like `default~`, some other places don't be so though.

Reply:
Agree, make them as const variables of this `planpreview` package is LGTM 👍

Reply:
Sure. Applied this. 👍